src/Particle/LongRange/StructFact.cpp

```diff
@@ -120,13 +115,23 @@ void StructFact::computeRhok(const ParticleSet& P)
         rhok_i_ptr[ki] += s;
       }
 #else
-      for (int ki = 0; ki < nk; ki++)
-        phiV[ki] = dot(k_lists_.kpts_cart[ki], pos);
-      eval_e2iphi(nk, phiV.data(), eikr_r_temp.data(), eikr_i_temp.data());
-      for (int ki = 0; ki < nk; ki++)
+      // make the compute over nk by blocks
+      constexpr size_t kblock_size = 512;
```

Comment:
I made this change to make phiV, eikr_r_temp, eikr_i_temp fixed size at compile time and lives on the stack.
On CPU this is usually called cache blocking and I will port to GPU similarly later.

Reply:
Iff this is profiled to still be a large kernel after offloading, I do wonder if the code inside the loop will need a bit more simplification to gain max benefit.

Reply:
It goes to GPU if offload is on. On CPU, I did see performance gain due to cache blocking. The biggest gain to me is getting rid of phiV, eikr_r_temp, eikr_i_temp as class members.